### PR TITLE
[#347] OpenAPI v3: validation for SecurityRequirement#{securityScheme…

### PR DESCRIPTION
**Pull request by ghillairet**
…Name}

See #347

### CODE REVIEW

File: com.reprezen.swagedit.openapi3.tests/src/com/reprezen/swagedit/openapi3/validation/ValidatorTest.xtend [hunk 1]
```diff
@@ -560,5 +558,133 @@ class ValidatorTest {
...
+		    get:
+		      security:
+		        - oauth: []
+		      responses:
+		        '200':
+		          description: OK
+		components:
+		  securitySchemes:
+		    oauth:
+		      type: http
```
Comment (tfesenko) [inline com.reprezen.swagedit.openapi3.tests/src/com/reprezen/swagedit/openapi3/validation/ValidatorTest.xtend]: `oauth` as _a little bit_ confusing name for an HTTP security scheme ;)

File: com.reprezen.swagedit.openapi3/src/com/reprezen/swagedit/openapi3/validation/OpenApi3Validator.java [hunk 2]
```diff
@@ -42,19 +43,74 @@ private void validateSecuritySchemeReferences(Model model, AbstractNode node, Se
...
+                        validateSecuritySchemScopes(node, field, securityScheme, errors);
                     }
                 }
             }
         }
     }
 
+    private List<String> oauthScopes = Lists.newArrayList("oauth2", "openIdConnect");
+
+    private void validateSecuritySchemScopes(AbstractNode node, String name, AbstractNode securityScheme,
```
Comment (tfesenko) [inline com.reprezen.swagedit.openapi3/src/com/reprezen/swagedit/openapi3/validation/OpenApi3Validator.java]: A typo: SecuritySchem**e**

File: com.reprezen.swagedit.openapi3/src/com/reprezen/swagedit/openapi3/validation/OpenApi3Validator.java [hunk 3]
```diff
@@ -42,19 +43,74 @@ private void validateSecuritySchemeReferences(Model model, AbstractNode node, Se
 
             if (node.isObject()) {
                 for (String field : node.asObject().fieldNames()) {
-                    if (securitySchemes.get(field) == null) {
+                    AbstractNode securityScheme = securitySchemes.get(field);
+
+                    if (securityScheme == null) {
                         errors.add(
                                 error(node.get(field), IMarker.SEVERITY_ERROR, Messages.error_invalid_reference_type));
+                    } else {
```
Comment (tfesenko) [inline com.reprezen.swagedit.openapi3/src/com/reprezen/swagedit/openapi3/validation/OpenApi3Validator.java]: I think here (on line 50), we have enough info about the expected type, so we can add an explanation at the end of the standard "Invalid object reference, the referenced object is not of expected type":
> Invalid object reference, the referenced object is not of expected type. It should be a valid security scheme.

File: com.reprezen.swagedit.core/src/com/reprezen/swagedit/core/validation/messages.properties [hunk 4]
```diff
@@ -42,6 +42,10 @@ content_assist_proposal_workspace = Press '%s' to show all %s in the workspace.
 outline_proposal_local = Press '%s' to show elements only the from current file
 outline_proposal_project = Press '%s' to show elements from the current project files
 outline_proposal_workspace = Press '%s' to show elements from the current workspace files
+error_scope_should_be_empty = Security schemes other than `oauth` or `openIdConnect` should have for requirement an empty array
+error_scope_should_not_be_empty = Security schemes `oauth` or `openIdConnect` should specify a list of scope names required for the execution
+error_invalid_scope_reference = Security requirement does not match any scope name as defined in the security scheme 
+
```
Comment (tfesenko) [inline com.reprezen.swagedit.core/src/com/reprezen/swagedit/core/validation/messages.properties]: @tedepstein , we need your input on phrasing for these error messages:

> Security schemes other than `oauth` or `openIdConnect` should have for requirement an empty array

> Security schemes `oauth` or `openIdConnect` should specify a list of scope names required for the execution

> Security requirement does not match any scope name as defined in the security scheme

File: com.reprezen.swagedit.openapi3/src/com/reprezen/swagedit/openapi3/validation/OpenApi3Validator.java [hunk 5]
```diff
@@ -42,19 +43,74 @@ private void validateSecuritySchemeReferences(Model model, AbstractNode node, Se
...
+    private List<String> oauthScopes = Lists.newArrayList("oauth2", "openIdConnect");
+
+    private void validateSecuritySchemScopes(AbstractNode node, String name, AbstractNode securityScheme,
+            Set<SwaggerError> errors) {
+        AbstractNode type = securityScheme.get("type");
+        if (type == null) {
+            return;
+        }
+
+        boolean hasScopes = oauthScopes.contains(type.asValue().getValue());
```
Comment (tfesenko) [inline com.reprezen.swagedit.openapi3/src/com/reprezen/swagedit/openapi3/validation/OpenApi3Validator.java]: `hasScopes` sounds like the node already has defined scope, while it's more like `shouldHaveScopes`

**Comment by tfesenko** [comment]
@ghillairet , thanks for a nice set of tests! Passed initial code review: 
* Requested several minor changes
* Possibly we will need to rephrase some of the new error messages

File: com.reprezen.swagedit.core/src/com/reprezen/swagedit/core/validation/messages.properties (hunk 4 above)
Comment (tedepstein) [reply com.reprezen.swagedit.core/src/com/reprezen/swagedit/core/validation/messages.properties]: @tfesenko , here's what I'd recomment:

> {schemeName} is defined as an {schemeType}-typed security scheme. Security requirements for scheme types other than oauth2 or openIdConnect must have an explicit empty array as the value.  Example: "{schemename} : []"

> {schemeName} is defined as an {schemeType}-typed security scheme. Security requirements for oauth or openIdConnect schemes must specify a list of scope names required for execution. 

> "{scopeName}" does not match any scope name defined in the {schemeName} security scheme.

**Comment by ghillairet** [comment]
@tfesenko 

I updated the messages and other suggestions.

Here are the new messages preceded by the invalid specs:

```yaml
  /pets/{petId}:
    get:
      summary: Info for a specific pet
      operationId: showPetById
      security:
        - testScheme:
           - foo          
      responses:
        '200':
          description: Expected response to a valid request
components:            
  securitySchemes:
    testScheme:
      type: http
```

screen shot 2017-10-06 at 17 50 01


```yaml
  /pets/{petId}:
    get:
      summary: Info for a specific pet
      operationId: showPetById
      security:
        - testScheme: [] 
      responses:
        '200':
          description: Expected response to a valid request
components:            
  securitySchemes:
    testScheme:
      type: oauth2
```

screen shot 2017-10-06 at 17 51 11


```yaml
  /pets/{petId}:
    get:
      summary: Info for a specific pet
      operationId: showPetById
      security:
        - testScheme:
          - foo
      responses:
        '200':
          description: Expected response to a valid request
components:            
  securitySchemes:
    testScheme:
      type: oauth2
      flows:
        implicit:
          scopes:
            "write:pets": something
```

screen shot 2017-10-06 at 17 52 11

**Comment by tedepstein** [comment]
Looks good, @ghillairet , thanks.

File: com.reprezen.swagedit.openapi3.tests/src/com/reprezen/swagedit/openapi3/validation/ValidatorTest.xtend [hunk 6]
```diff
@@ -530,161 +527,182 @@ class ValidatorTest {
...
+			  /foo/{bar}:
+			    get:
+			      security:
+			        - basic: []
+			      responses:
+			        '200':
+			          description: OK
+			components:
+			  securitySchemes:
+			    basic:
```
Comment (tfesenko) [inline com.reprezen.swagedit.openapi3.tests/src/com/reprezen/swagedit/openapi3/validation/ValidatorTest.xtend:658]: :)

File: com.reprezen.swagedit.openapi3/src/com/reprezen/swagedit/openapi3/validation/OpenApi3Validator.java [hunk 7]
```diff
@@ -46,10 +46,12 @@ private void validateSecuritySchemeReferences(Model model, AbstractNode node, Se
                     AbstractNode securityScheme = securitySchemes.get(field);
 
                     if (securityScheme == null) {
-                        errors.add(
-                                error(node.get(field), IMarker.SEVERITY_ERROR, Messages.error_invalid_reference_type));
+                        String message = Messages.error_invalid_reference_type
+                                + " It should be a valid security scheme.";
```
Comment (tfesenko) [inline com.reprezen.swagedit.openapi3/src/com/reprezen/swagedit/openapi3/validation/OpenApi3Validator.java]: Thanks!

**Comment by tfesenko** [comment]
Thanks! Passed QA and code review

**Pull request merged**